Modularize the Color Picker via properties and new picker mode.

cridenour opened the pull request:
As noted in #67663, the new color picker removed some flexibility. The new options are great, but limited how we could use the control in game. 

Inspired by this, I took a stab at giving full flexibility in the color picker. Properties are exposed to hide all elements outside the picker itself, which can also be disabled with a new Picker Mode. 

The existing `presets_visible` property was no longer respected (launching with it enabled still hid the presets until Swatches was clicked) so I used that to hide the Swatches and Recent Colors buttons.

Similarly, after the new UX `presets_enabled` worked correctly, but was confusing as internally both recent colors and swatches are considered presets. I changed the name of this property to `can_add_swatches`. 

The picker itself is a fantastic control and the hope is to make it helpful in more scenarios with additional customization options.

*Bugsquad edit:*
- Fixes #67663

cridenour committed:
Modularize the Color Picker via properties.

RedMser commented:
Seems like the XML documentation is missing the `SHAPE_NONE` constant.

cridenour commented:
> Seems like the XML documentation is missing the `SHAPE_NONE` constant.

Thanks for the catch. Updated.

cridenour commented:
New push to fix property ordering (actually using `doctool` this time)

KoBeWi commented:
I tested this and it's overall good. But I have a few notes:
- all `something_visible` properties should be inside a group, e.g. "Customization"
- `modes_visible` should be `color_modes_visible`
- I wonder if `can_add_swatches` is even necessary. I know it was there, but it just disables a button. Kind of pointless if you can just disable presets altogether. Up to discussion I guess
- it would be nice if the properties were ordered in the order of elements they represent. Right now it seems random

KoBeWi commented on doc/classes/ColorPicker.xml:
```diff
@@ -55,6 +55,9 @@
 	</methods>
 	<members>
+		<member name="can_add_swatches" type="bool" setter="set_can_add_swatches" getter="are_swatches_enabled" default="true">
+			If [code]true[/code], the "add swatch" button is enabled.
```
That button isn't called "add swatch". Maybe something like
```suggestion
			If [code]true[/code], it's possible to add presets under Swatches.
```
would be better.

KoBeWi commented on doc/classes/ColorPicker.xml:
```diff
@@ -67,14 +70,23 @@
 			If [code]true[/code], shows an alpha channel slider (opacity).
 		</member>
+		<member name="hex_visible" type="bool" setter="set_hex_visible" getter="is_hex_visible" default="true">
+			If [code]true[/code], the hex input is visible.
```
```suggestion
			If [code]true[/code], the hex input field is visible.
```

KoBeWi commented on doc/classes/ColorPicker.xml:
```diff
@@ -67,14 +70,23 @@
+			If [code]true[/code], the hex input is visible.
+		</member>
+		<member name="modes_visible" type="bool" setter="set_modes_visible" getter="are_modes_visible" default="true">
+			If [code]true[/code], the RGB / HSV / RAW selector is visible.
```
```suggestion
			If [code]true[/code], the color mode buttons visible.
```
It's better to not mention them by name, as the list might change (also there is `...`  button that the description did not include).

KoBeWi commented on doc/classes/ColorPicker.xml:
```diff
@@ -67,14 +70,23 @@
+			If [code]true[/code], the Swatches and Recent Colors presets are visible.
+		</member>
+		<member name="sampler_visible" type="bool" setter="set_sampler_visible" getter="is_sampler_visible" default="true">
+			If [code]true[/code], the color sampler is visible.
```
```suggestion
			If [code]true[/code], the color sampler and color preview are visible.
```
I wonder if it sampler and preview shouldn't be split further. They are technically independent, just come in the same row.

KoBeWi commented on doc/classes/ColorPicker.xml:
```diff
@@ -125,6 +137,9 @@
 			HSL OK Color Model circle color space.
 		</constant>
+		<constant name="SHAPE_NONE" value="4" enum="PickerShapeType">
+			Hides the color picker.
```
This description is very misleading. Better:
```suggestion
			The color space shape is hidden.
```
(consistent with `picker_shape` property)

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -607,10 +616,18 @@ void ColorPicker::set_picker_shape(PickerShapeType p_shape) {
-	shape_popup->set_item_checked(current_shape, false);
-	shape_popup->set_item_checked(p_shape, true);
+	if (current_shape != SHAPE_NONE) {
+		shape_popup->set_item_checked(current_shape, false);
```
Does this affect anything? If the shape button is hidden, the popup can't be opened, so it doesn't really matter what it has selected.

cridenour replied:
I was getting index errors since the popup item didn't include an entry for the none shape.

KoBeWi replied:
Ah, right..

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -607,10 +616,18 @@ void ColorPicker::set_picker_shape(PickerShapeType p_shape) {
+	}
 
-	btn_shape->set_icon(shape_popup->get_item_icon(p_shape));
+	if (p_shape == SHAPE_NONE) {
+		btn_shape->set_icon(shape_popup->get_item_icon(SHAPE_HSV_RECTANGLE));
+	} else {
+		btn_shape->set_icon(shape_popup->get_item_icon(p_shape));
+	}
```
```suggestion
	if (p_shape != SHAPE_NONE) {
		btn_shape->set_icon(shape_popup->get_item_icon(p_shape));
	}
```
Simpler. Icon is irrelevant if the button is invisible.

KoBeWi commented on scene/gui/color_picker.h:
```diff
@@ -125,6 +126,10 @@ class ColorPicker : public BoxContainer {
 	PopupMenu *shape_popup = nullptr;
 	PopupMenu *mode_popup = nullptr;
 	MenuButton *btn_shape = nullptr;
+	HBoxContainer *mode_hbc = nullptr;
+	HBoxContainer *hb_smpl = nullptr;
+	GridContainer *slider_gc = nullptr;
+	HBoxContainer *hex_hb = nullptr;
```
I think these could get better names if they are no longer local.

cridenour replied:
I'm happy to leave things better than how I found them - can you point me to a node that has properly named variables to model after?

KoBeWi replied:
I think it's enough if you rename `hb_smpl` to `sample_hb`, to at least be consistent with naming.

cridenour commented:
Updated per @KoBeWi's review: 

- Added a Customization group for the properties.
- Re-ordered properties to match the visually flow of the control
- Renamed `modes_visible` to `color_modes_visible`
- Updated documentation wording to suggested

Outstanding:
 - Resolution on open comments
 - Discussion about `can_add_swatches` and the utility it provides.

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -606,10 +615,16 @@ void ColorPicker::set_picker_shape(PickerShapeType p_shape) {
+		shape_popup->set_item_checked(current_shape, false);
+	}
+	if (p_shape != SHAPE_NONE) {
+		shape_popup->set_item_checked(p_shape, true);
+	}
 
-	btn_shape->set_icon(shape_popup->get_item_icon(p_shape));
+	if (p_shape != SHAPE_NONE) {
```
Same condition twice. You can move the code under 1 `if` statement.

KoBeWi commented on doc/classes/ColorPicker.xml:
```diff
@@ -55,26 +55,38 @@
 			The currently selected color mode. See [enum ColorModeType].
 		</member>
+		<member name="color_modes_visible" type="bool" setter="set_modes_visible" getter="are_modes_visible" default="true">
+			If [code]true[/code], the color mode buttons visible.
```
```suggestion
			If [code]true[/code], the color mode buttons are visible.
```

KoBeWi reviewed (approved):
Left some minor comments, but otherwise looks ok.

As for the `can_add_swatches` property, its removal can still be discussed and done in another PR. But with the changes here, it's possible to make a ColorPicker that contains only swatches, so disabling the button *might* make sense. But maybe it would be better if the property was hiding it, idk

cridenour commented:
Thanks @KoBeWi! All outstanding comments have been resolved.

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:
